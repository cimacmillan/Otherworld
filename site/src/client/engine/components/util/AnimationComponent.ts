import { CompositeAnimation } from "../../../util/animation/CompositeAnimation";
import { GameAnimation } from "../../../util/animation/GameAnimation";
import { Entity } from "../../Entity";
import { EntityComponent } from "../../EntityComponent";

type AnimationCreationFunction<T> = (
    entity: Entity<T>
) => GameAnimation | CompositeAnimation;

export function AnimationComponent<T>(
    createAnimation: AnimationCreationFunction<T>
): EntityComponent<T> {
    let animation: GameAnimation | CompositeAnimation;
    return {
        getActions: (entity: Entity<T>) => ({
            onEntityCreated: () => {
                animation = createAnimation(entity);
                animation.start();
            },
            onEntityDeleted: () => animation.stop(),
        }),
        update: () => animation.tick(),
    };
}
