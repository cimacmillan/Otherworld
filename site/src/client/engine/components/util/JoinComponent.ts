import { Actions, emptyActions } from "../../../Actions";
import { FunctionEventSubscriber } from "@cimacmillan/refunc";
import { Entity } from "../../Entity";
import { EntityComponent } from "../../EntityComponent";

export function JoinComponent<T>(
    components: Array<EntityComponent<Partial<T>>>
): EntityComponent<T> {
    const subscriber = new FunctionEventSubscriber<Actions>(emptyActions);
    let initialised = false;
    return {
        update: (entity: Entity<T>) =>
            components.forEach(
                (component) => component.update && component.update(entity)
            ),
        getActions: (entity: Entity<T>) => {
            if (!initialised) {
                initialised = true;
                components.forEach((component) =>
                    subscriber.subscribe(component.getActions(entity))
                );
            }
            return subscriber.actions();
        },
    };
}
