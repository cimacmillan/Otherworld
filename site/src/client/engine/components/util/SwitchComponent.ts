import { Actions, emptyActions } from "../../../Actions";
import { FunctionEventSubscriber } from "@cimacmillan/refunc";
import { Entity } from "../../Entity";
import { EntityComponent } from "../../EntityComponent";

export interface SwitchComponents {
    [key: string]: EntityComponent<any>;
}

export class SwitchComponent<T> implements EntityComponent<T> {
    private newState: string;

    constructor(
        private components: SwitchComponents,
        private currentState: string,
        private getSwitch: (entity: Entity<T>) => string
    ) {
        this.newState = currentState;
    }

    public update(entity: Entity<T>) {
        if (this.newState !== this.currentState) {
            const currentStateCallback = this.components[this.currentState]
                ? this.components[this.currentState].getActions(entity)
                : {};
            const newStateCallback = this.components[this.newState]
                ? this.components[this.newState].getActions(entity)
                : {};

            currentStateCallback.onEntityDeleted &&
                currentStateCallback.onEntityDeleted();
            newStateCallback.onEntityCreated &&
                newStateCallback.onEntityCreated();

            this.currentState = this.newState;
        }
        const currentStateCallback = this.components[this.currentState];

        currentStateCallback &&
            currentStateCallback.update &&
            currentStateCallback.update(entity);
    }

    public getActions(entity: Entity<T>) {
        const subscriber = new FunctionEventSubscriber<Actions>(emptyActions);
        Object.keys(this.components).forEach((key) => {
            subscriber.subscribe(
                this.wrapActions(
                    this.components[key].getActions(entity),
                    key,
                    entity
                )
            );
        });

        const entityActions = subscriber.actions();
        return {
            ...entityActions,
            onStateTransition: (from: any, to: any) => {
                entityActions.onStateTransition &&
                    entityActions.onStateTransition(from, to);
                this.newState = this.getSwitch(entity);
            },
        };
    }

    private wrapActions(
        actions: Partial<Actions>,
        rightKey: string,
        entity: Entity<T>
    ): Partial<Actions> {
        const newActions: Partial<Actions> = {};
        Object.keys(actions).forEach((key) => {
            newActions[key as keyof Actions] = (...args: any[]) => {
                if (this.getSwitch(entity) === rightKey) {
                    const action = actions[key as keyof Actions];
                    (action as any)(...args);
                }
            };
        });
        return newActions;
    }
}
