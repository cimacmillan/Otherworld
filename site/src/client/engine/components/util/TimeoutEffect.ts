import { ProcedureService } from "../../../services/jobs/ProcedureService";
import { Entity } from "../../Entity";
import { EntityComponent } from "../../EntityComponent";

export function TimeoutComponent<T>(
    callback: (entity: Entity<T>) => void,
    time: number,
    onPrematureQuit?: () => void
): EntityComponent<T> {
    let timeout: number;
    let done: boolean = false;
    return {
        getActions: (entity: Entity<T>) => ({
            onEntityCreated: () => {
                timeout = ProcedureService.setGameTimeout(() => {
                    done = true;
                    callback(entity);
                }, time);
            },
            onEntityDeleted: () => {
                ProcedureService.clearTimeout(timeout);
                if (done === false) {
                    onPrematureQuit && onPrematureQuit();
                }
            },
        }),
    };
}
