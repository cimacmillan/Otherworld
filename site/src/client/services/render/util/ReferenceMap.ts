import { vec3 } from "gl-matrix";



export class ReferenceMap<T> {
    private map: T[] = [];

    public constructor(
        private isEqual: (a: T, b: T) => boolean
    ) {

    } 

    public getMap(): T[] {
        return this.map;
    }

    public getID(a: T): number {
        const id = this.map.findIndex((value: T, index: number) => this.isEqual(value, a));
        if (id >= 0) {
            return id;
        }
        const newId = this.map.push(a);
        return (newId - 1);
    }
}

export class ColourMap extends ReferenceMap<vec3> {
    public constructor() {
        super((a: vec3, b: vec3) => {
            return a[0] === b[0] && a[1] === b[1] && a[2] === b[2]
        })
    }
}

