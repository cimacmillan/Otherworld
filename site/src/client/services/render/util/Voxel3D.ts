import { ServiceLocator } from "../../ServiceLocator";
import { mat4, vec3 } from "gl-matrix";
import { RenderItem } from "../types/RenderInterface";
import { Object3D } from "../services/ObjectRenderService";

interface VoxelData {
    position: vec3;
    size: number;
    colour: vec3;
    transform?: mat4;
}

const IDENTITY = mat4.identity(mat4.create());

export class Voxel3D {
    private item?: RenderItem;
    private itemData?: Object3D;

    public constructor (private serviceLocator: ServiceLocator, private data: VoxelData) {
    }

    public setColour(colour: vec3) {
        this.data.colour = colour;
        if (this.item && this.itemData) {
            this.serviceLocator.getRenderService().triangleRenderService.updateItem(this.item, {
                colour: this.itemData.colour.map(() => colour)
            })
        }
    }
    
    public attach() {
        if (this.item) {
            return;
        }
        const { triangleRenderService } = this.serviceLocator.getRenderService();
        const { position, size, colour } = this.data;
        const transform = this.data.transform || IDENTITY;

        this.itemData = {
            positions: [],
            colour: [],
            transform
        }

        const createTriangle = (positions: [vec3, vec3, vec3], colour: vec3) => {
            this.itemData.positions.push(positions);
            this.itemData.colour.push(colour);
        }

        const createQuad = (positions: [vec3, vec3, vec3, vec3], col: vec3) => {
            createTriangle([positions[0], positions[1], positions[2]], col);
            createTriangle([positions[0], positions[3], positions[2]], col);
        }

        const [x, y, z] = position;


        const colourVariance = 0.4;
        const darker = vec3.scale(vec3.create(), colour, 1 - (colourVariance / 2));
        const darkest = vec3.scale(vec3.create(), colour, 1 - (colourVariance));

        //Back
        createQuad([
            [x, y, z],
            [x + size, y, z],
            [x + size, y + size, z],
            [x, y + size, z]
        ], colour
        );

        //Front
        createQuad([
            [x, y, z + size],
            [x + size, y, z + size],
            [x + size, y + size, z + size],
            [x, y + size, z + size]
        ], colour);

        //Left
        createQuad([
            [x, y, z],
            [x, y, z + size],
            [x, y + size, z + size],
            [x, y + size, z]
        ], darkest);

        //Right
        createQuad([
            [x + size, y, z],
            [x + size, y, z + size],
            [x + size, y + size, z + size],
            [x + size, y + size, z]
        ], darkest);

        //Bottom
        createQuad([
            [x, y, z],
            [x + size, y, z],
            [x + size, y, z + size],
            [x, y, z + size]
        ], darker);

        //Top
        createQuad([
            [x, y + size, z],
            [x + size, y + size, z],
            [x + size, y + size, z + size],
            [x, y + size, z + size]
        ], darker);

        this.item = triangleRenderService.createItem(this.itemData);
    }

    public destroy() {
        const { triangleRenderService } = this.serviceLocator.getRenderService();
        if (this.item) {
            triangleRenderService.freeItem(this.item);
        }
        this.item = undefined;
    }

    public isBeingDrawn() {
        return !!this.item;
    }
}
