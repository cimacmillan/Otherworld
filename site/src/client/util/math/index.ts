export * from "./SyncedArray";
export * from "./Basic";
export * from "./Sorting";
export * from "./Vector";
export * from "./Random";
