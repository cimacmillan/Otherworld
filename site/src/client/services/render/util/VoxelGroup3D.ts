import { mat4, vec3 } from "gl-matrix";
import { forEach3D } from "../../../util/math";
import { ServiceLocator } from "../../ServiceLocator";
import { Object3D } from "../services/ObjectRenderService";
import { RenderItem } from "../types/RenderInterface";
import { Voxel3D } from "./Voxel3D";

export interface VoxelGroupData {
    data: number[][][],
    colourMap: vec3[],
    sizePerVoxel: number;
}

export class VoxelGroup3D {
    private voxels: Voxel3D[][][];
    private transform: mat4 = mat4.identity(mat4.create());
    private position: vec3 = [0, 0, 0];
    private angle: vec3 = [0, 0, 0];
    private center: vec3 = [0, 0, 0];
    private scale: vec3 = [1, 1, 1];
    private renderItem?: [RenderItem, Object3D];

    public constructor(private serviceLocator: ServiceLocator, private voxelGroupData: VoxelGroupData) {
        const { data, colourMap, sizePerVoxel } = voxelGroupData;

        this.voxels = data.map(
            (ys, x) => ys.map(
                (zs, y) => zs.map(
                    (voxelId, z) => {
                        const voxel = new Voxel3D(serviceLocator, {
                            position: [
                                x * sizePerVoxel,
                                y * sizePerVoxel,
                                z * sizePerVoxel
                            ],
                            size: sizePerVoxel,
                            colour: colourMap[voxelId ? voxelId - 1 : 0]
                        });
                        return voxel;
                    }
                )
            )
        )
    }

    public attach() {
        this.onDataUpdate();
    }

    public destroy() {
        forEach3D(this.voxels, (voxel: Voxel3D) => {
            voxel.destroy();
        });
        if (this.renderItem) {
            this.serviceLocator.getRenderService().triangleRenderService.freeItem(this.renderItem[0]);
            this.renderItem = undefined;
        }
    }

    public onDataUpdate() {
        const { colourMap, data } = this.voxelGroupData;

        const object3D: Object3D = {
            positions: [],
            colour: [],
            transform: this.transform
        }

        forEach3D(this.voxels, (voxel: Voxel3D, x: number, y: number, z: number) => {
            const voxelId = data[x][y][z];
            if (voxelId > 0) {
                voxel.setColour(colourMap[voxelId - 1]);
                voxel.attach();
            } else {
                voxel.destroy();
            }

            const itemData = voxel.getData();
            itemData && object3D.positions.push(...itemData.positions);
            itemData && object3D.colour.push(...itemData.colour);
        });

        if (this.renderItem) {

            const [item, object] = this.renderItem;
            
            if (object3D.positions.length === object.positions.length) {
                this.serviceLocator.getRenderService().triangleRenderService.updateItem(item, object3D);
            } else {
                this.serviceLocator.getRenderService().triangleRenderService.freeItem(item);
                const newItem = this.serviceLocator.getRenderService().triangleRenderService.createItem(object3D);
                this.renderItem = [newItem, object3D];
            }

        } else {
            const item = this.serviceLocator.getRenderService().triangleRenderService.createItem(object3D);
            this.renderItem = [item, object3D];
        }
    } 

    public setPosition(position: vec3) {
        this.position = position;
        this.recomputeTransform();
    }

    public setAngle(angle: vec3) {
        this.angle = angle;
        this.recomputeTransform();
    }

    public setCenter(center: vec3) {
        this.center = center;
        this.recomputeTransform();
    }

    public setScale(scale: vec3) {
        this.scale = scale;
        this.recomputeTransform();
    }

    private recomputeTransform() {
        mat4.identity(this.transform);    
        mat4.translate(this.transform, this.transform, this.position);
        mat4.translate(this.transform, this.transform, [this.center[0], this.center[1], this.center[2]]);
        mat4.rotateX(this.transform, this.transform, this.angle[0]);
        mat4.rotateY(this.transform, this.transform, this.angle[1]);
        mat4.rotateZ(this.transform, this.transform, this.angle[2]);
        mat4.translate(this.transform, this.transform, [-this.center[0], -this.center[1], -this.center[2]]);
    }

}

