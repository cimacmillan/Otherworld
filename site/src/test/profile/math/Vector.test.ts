import { vec_add, vec_add_mutable, vec_interpolate, vec_sum, vec_sum_immutable } from "../../../client/util/math";
import { profile, SAMPLE_BIG, SAMPLE_MED } from "../ProfileUnit";

describe("Vector Profile", () => {
    test("add", () => {
        const immutableA = {x: 1, y: 2};
        const immutableB = {x: 2, y: 3};
        profile(SAMPLE_BIG, () => vec_add(immutableA, immutableB));
    });

    test("add_mutable", () => {
        const mutableA = {x: 1, y: 2};
        const mutableB = {x: 2, y: 3};
        profile(SAMPLE_BIG, () => vec_add_mutable(mutableA, mutableB));
    });

    test("vec_sum", () => {
        const vecs = [{x: 1, y: 2}, {x: 3, y: 4}, {x: 5, y: 6}];
        profile(SAMPLE_MED, () => vec_sum(vecs));
    });

    test("vec_sum_immutable", () => {
        const vecs = [{x: 1, y: 2}, {x: 3, y: 4}, {x: 5, y: 6}];
        profile(SAMPLE_MED, () => vec_sum_immutable(vecs));
    });

    test("vec_interpolate", () => {
        const vecs = [{x: 1, y: 2}, {x: 3, y: 4}, {x: 5, y: 6}];
        const alphas = [0.1, 0.2, 0.3];
        profile(SAMPLE_MED, () => vec_interpolate(vecs, alphas));
    });

});
