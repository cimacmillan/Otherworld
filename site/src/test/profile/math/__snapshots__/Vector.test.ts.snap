// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vector Profile add 1`] = `1036`;

exports[`Vector Profile add_mutable 1`] = `10859`;

exports[`Vector Profile vec_interpolate 1`] = `57`;

exports[`Vector Profile vec_sum 1`] = `45`;

exports[`Vector Profile vec_sum_immutable 1`] = `55`;
